File: binary-tree-inorder-traversal-94.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

// class Solution {
// public:
//   vector<int> inorderTraversal(TreeNode *root) {
//     TreeNode *curr = root;
//     stack<TreeNode *> s;
//     vector<int> result;
//     while (1) {
//       while (curr) {
//         s.push(curr);
//         curr = curr->left;
//       }
//       if (s.empty()) {break;}
//       curr = s.top();
//       s.pop();
//       result.push_back(curr->val);
//       curr = curr->right;
//     }
//     return result;
//   }
// };
class Solution {
public:
  vector<int> inorderTraversal(TreeNode *root) {
    TreeNode *curr = root;
    stack<TreeNode *> s;
    vector<int> result;
    while (curr || !s.empty()) {
      if (curr) {
        s.push(curr);
        curr = curr->left;
      } else {
        curr = s.top();
        s.pop();
        result.push_back(curr->val);
        curr = curr->right;
      }
    }
    return result;
  }
};

File: binary-tree-inorder-traversal-94.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	var res []int
	var stack []*TreeNode
	curr := root
	for len(stack) > 0 || curr != nil {
		if curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		} else {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res = append(res, node.Val)
			if node.Right != nil {
				curr = node.Right
			}
		}
	}
	return res
}
